### PR TITLE
Logic for time_zone sys to local conversion

### PR DESCRIPTION
**Pull request by d-winsor**
This contains the logic for getting `time_zone` transition information from the ICU and using that to convert between `sys_time<>` and `local_time<>` according to the rules of the specified `time_zone`.

Glancing over the tests first may be useful in understanding the different corner cases that need to be considered, particularly around ambiguous and nonexistent `local_time` conversions.

### CODE REVIEW

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 1]
```diff
@@ -80,6 +92,20 @@ void timezone_names_test() {
     assert(link3 <=> link1 == strong_ordering::greater);
 #endif // __cpp_lib_concepts
 
+    try {
+        // ensure locate_zone returns time_zone with given name
+        assert(all_of(tzdb.zones.begin(), tzdb.zones.end(),
+            [&](const auto& zone) { return tzdb.locate_zone(zone.name())->name() == zone.name(); }));
+        // ensure locate_zone returns target of time_zone
```
Comment (d-winsor) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:99]:
```suggestion
        // ensure locate_zone returns correct target of time_zone_link

```

File: stl/inc/chrono [hunk 2]
```diff
@@ -2229,7 +2270,140 @@ namespace chrono {
...
+        template <class _Duration>
+        _NODISCARD local_info get_info(const local_time<_Duration>& _Local) const {
+            local_info _Info{};
+            _Info.first = _Get_info(_Local.time_since_epoch());
+
+            const sys_seconds _Local_sys{duration_cast<sys_seconds::duration>(_Local.time_since_epoch())};
+            const auto _Curr_sys = _Local_sys - _Info.first.offset;
+            if (_Info.first.begin != _Min_seconds && _Curr_sys < _Info.first.begin + days{1}) {
+                // get previous transition information
+                _Info.second = get_info(_Info.first.begin - seconds{1});
```
Comment (d-winsor) [inline stl/inc/chrono:2287]: Calling this twice does incur an extra cost of transforming a string and reloading a ICU calendar. An alternative solution is to have a new `__std__tzdb_get_local_info` function and do all this logic there. 

I originally had that solution but it added a lot more code complexity, particularly around allocations and error handling. In my mind the slight performance gain doesn't justify this corner case (happens 4/256 times because there is generally 2 transitions a year).

Comment (d-winsor) [reply stl/inc/chrono:2287]: If this is the desired change lets scope it to a future PR in `feature/chrono`

File: stl/src/tzdb.cpp [hunk 3]
```diff
@@ -213,17 +264,62 @@ namespace {
...
+        _Icu_get_string_fn _Icu_fn = [_Cal, _Display_type](
+                                         UChar* _Result, int32_t _Result_capacity, UErrorCode* _UErr) {
+            return __icu_ucal_getTimeZoneDisplayName(_Cal, _Display_type, nullptr, _Result, _Result_capacity, _UErr);
+        };
+        return _Get_icu_string_impl(_Icu_fn, 12, _Result_len, _Err);
+    }
+
+    _STD unique_ptr<UCalendar, decltype(&__icu_ucal_close)> _Get_cal(
+        const char* _Tz, const size_t _Tz_len, __std_tzdb_error& _Err) noexcept {
+        const auto _Tz_name = _Allocate_narrow_to_wide(_Tz, _Tz_len, _Err);
```
Comment (d-winsor) [inline stl/src/tzdb.cpp:300]:
```suggestion
        const auto _Tz_name = _Allocate_narrow_to_wide(_Tz, static_cast<int>(_Tz_len), _Err);
```

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 4]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
     assert(_Locate_zone_impl(tzdb.zones, "America/Nuuk") == nullptr); // does not exist in ICU (very rare)
 }
 
+void validate_timezone_transitions(const time_zone* tz, const Transition& data) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:120]: `data` is the name of a function in `std`; I recommend choosing any other name to avoid "shadowing" even if there are no shadow warnings.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 5]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
...
+    auto info = tz->get_info(data.begin());
+    assert(info.begin == data.begin());
+    assert(info.end == data.end());
+    assert(info.offset == data.offset());
+    assert(info.save == data.save());
+    assert(info.abbrev == data.abbrev());
+}
+
+void timezone_sys_info_test() {
+    const auto& tzdb = get_tzdb();
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:130]: `tzdb` is the name of a `std::chrono` type; I strongly recommend another name for this reference. Occurs below.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 6]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
...
+
+    auto la_tz = tzdb.locate_zone(LA::Tz_name);
+    assert(la_tz != nullptr);
+    validate_timezone_transitions(la_tz, LA::Day_1);
+    validate_timezone_transitions(la_tz, LA::Std_1);
+    validate_timezone_transitions(la_tz, LA::Day_2);
+
+    auto begin  = sydney_tz->get_info(Sydney::Std_1.begin());
+    auto middle = sydney_tz->get_info(Sydney::Std_1.begin() + days{1});
+    auto end    = sydney_tz->get_info(Sydney::Std_1.end());
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:146]: `begin` and `end` are the names of functions in `std`; I recommend choosing different names.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 7]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
...
+    // Only a single transition in UTC
+    assert(min_utc == max_utc);
+    assert(min_utc.begin < max_utc.end);
+    assert(min_utc.begin == time_zone::_Min_seconds);
+    assert(min_utc.end == time_zone::_Max_seconds);
+
+    // Test abbrevations other than standard/daylight savings such as war time.
+    // These senarios are not handled correctly by icu.dll
+    auto war_time = la_tz->get_info(sys_days{year{1942} / April / day{1}});
+    assert(war_time.abbrev == "PDT"); // IANA datbase == "PWT"
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:171]: Typo, should be "database".

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 8]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
...
+        assert(tz->to_local(to_standard - minutes{30}) == Day_to_Std.first.local_end() - minutes{30});
+
+        const auto& to_daylight = Std_to_Day.first.end();
+        assert(tz->to_local(to_daylight) == Std_to_Day.second.local_begin());
+        assert(tz->to_local(to_daylight + minutes{30}) == Std_to_Day.second.local_begin() + minutes{30});
+        assert(tz->to_local(to_daylight - minutes{30}) == Std_to_Day.first.local_end() - minutes{30});
+    }
+}
+
+void assert_local(const time_zone* tz, local_seconds time, int result, const sys_info& first, const sys_info& second) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:208]: `time` is the name of a CRT function.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 9]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
...
+    // time_zone::to_sys depends heavily on local_info so just test it here
+    // to exhaust all corner cases.
+    sys_seconds sys_earliest{time.time_since_epoch() - info.first.offset};
+    sys_seconds sys_latest{time.time_since_epoch() - info.second.offset};
+    try {
+        assert(tz->to_sys(time) == sys_earliest);
+        assert(result == local_info::unique);
+    } catch (nonexistent_local_time ex) {
+        assert(result == local_info::nonexistent);
+    } catch (ambiguous_local_time ex) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:223]: Should also be const ref, and should be unnamed (as `ex` is unused).

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 10]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
...
+    } else if (result == local_info::nonexistent) {
+        assert(tz->to_sys(time, choose::earliest) == info.first.end);
+        assert(tz->to_sys(time, choose::latest) == info.first.end);
+    } else if (result == local_info::ambiguous) {
+        assert(tz->to_sys(time, choose::earliest) == sys_earliest);
+        assert(tz->to_sys(time, choose::latest) == sys_latest);
+    }
+}
+
+void validate_get_local_info(const time_zone* tz, const std::pair<Transition, Transition>& transition, int result) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:239]: Should include `<utility>` for `pair`, and should be unqualified (this test is already `using namespace std;`).

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 11]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
...
+    sys_info empty{};
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:240]: `empty` is also the name of a function in `std`.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 12]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
...
+    // Get the local time for the beginning of the ambiguous/nonexistent section
+    const auto danger_begin = get_danger_begin(transition.first, transition.second);
+    const auto danger_end   = get_danger_end(transition.first, transition.second);
+    (void) result;
+    assert_local(tz, danger_begin - days{2}, local_info::unique, first, empty); // two days before
+    assert_local(tz, danger_begin - hours{1}, local_info::unique, first, empty); // one hour before
+    assert_local(tz, danger_begin, result, first, second); // danger begin
+    assert_local(tz, danger_begin + minutes{30}, result, first, second); // danger middle
+    assert_local(tz, danger_end, local_info::unique, second, empty); // danger end
+    assert_local(tz, danger_end + hours{2}, local_info::unique, second, empty); // one hour after
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:253]: Code says `hours{2}` but comment says `one hour`.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 13]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
...
+        // negative offset (UTC-8/-7) can fall in next transition
+        using namespace LA;
+        auto tz = tzdb.locate_zone(Tz_name);
+        assert(tz != nullptr);
+        validate_get_local_info(tz, Day_to_Std, local_info::ambiguous);
+        validate_get_local_info(tz, Std_to_Day, local_info::nonexistent);
+    }
+}
+
+template <typename _Dur>
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:277]: Should avoid `_Ugly` in test code.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 14]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
...
+    assert(tz->get_info(danger_end - precision).result != local_info::unique); // just inside danger zone
+    assert(tz->get_info(danger_begin).result != local_info::unique); // exact start of danger zone
+    assert(tz->get_info(danger_begin - precision).result == local_info::unique); // just before danger zone
+
+    // test precision is not lost when converting to local
+    assert(tz->to_local(transition) == second.local_begin<_Dur>());
+    assert(tz->to_local(transition + precision) == second.local_begin<_Dur>() + precision);
+    assert(tz->to_local(transition - precision) == first.local_end<_Dur>() - precision);
+
+    // test precision is not lost when converting to local
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:302]: This comment `converting to local` is repeated, but it looks like this test is `to_sys`.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 15]
```diff
@@ -91,9 +117,268 @@ void timezone_names_test() {
...
+        (void) tz->to_sys(danger_end - precision);
+        assert(false);
+    } catch (nonexistent_local_time&) {
+    } catch (ambiguous_local_time&) {
+    }
+}
+
+void timezone_precision_test() {
+    const auto& tzdb = get_tzdb();
+    using Milli_Dur  = duration<double, milli>;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:326]: Should technically include `<ratio>`.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/timezone_data.h [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+
+    constexpr string_view abbrev() const {
+        return _abbrev;
+    }
+
+    constexpr bool is_daylight() const {
+        return _save != minutes{0};
+    }
+
+    template <class _Duration = seconds>
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/timezone_data.h:38]: `_Ugly` identifier in test code. Occurs below.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/timezone_data.h [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+constexpr local_time<_Duration> get_danger_end(const Transition& first, const Transition& second) {
+    assert(first.end<_Duration>() == second.begin<_Duration>());
+    return second.local_begin<_Duration>() + first.save();
+}
+
+// Sydney
+// Standard time (AEST : UTC+10) -1 @ 3am
+// Daylight time (AEDT : UTC+11) +1 @ 2am
+namespace Sydney {
+    static constexpr std::string_view Tz_name{"Australia/Sydney"sv};
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/timezone_data.h:85]: This test is already `using namespace std;`. Occurs below.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/timezone_data.h [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+    static constexpr auto Daylight_begin_2020 =
+        sys_seconds{sys_days{year{2020} / October / day{4}}} + hours{2} - Standard_offset;
+    static constexpr auto Standard_begin_2021 =
+        sys_seconds{sys_days{year{2021} / April / day{4}}} + hours{3} - Daylight_offset;
+
+    Transition Day_1{Tz_name, Daylight_offset, hours{1}, Daylight_abbrev, Daylight_begin_2019, Standard_begin_2020};
+    Transition Std_1{Tz_name, Standard_offset, hours{0}, Standard_abbrev, Standard_begin_2020, Daylight_begin_2020};
+    Transition Day_2{Tz_name, Daylight_offset, hours{1}, Daylight_abbrev, Daylight_begin_2020, Standard_begin_2021};
+
+    std::pair<Transition, Transition> Day_to_Std{Day_1, Std_1};
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/timezone_data.h:103]: Should include `<utility>`.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/timezone_data.h (hunk 17 above)
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/timezone_data.h:85]: It doesn't really matter, but as namespace-scope constants in a header, these should ideally be `inline constexpr`, which fuses all definitions (even in separate translation units) together.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/timezone_data.h [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+    std::pair<Transition, Transition> Std_to_Day{Std_1, Day_2};
```
Comment (StephanTLavavej) [inline tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/timezone_data.h:104]: These are modifiable globals. I see that `Transition` stores `string _name;` but it's constructed from `string_view` and returns `string_view`. Would it be possible to store `string_view`, make `Transition`'s constructor `constexpr`, and make all of these `constexpr`?

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/chrono [hunk 20]
```diff
@@ -2229,7 +2270,140 @@ namespace chrono {
...
+        }
+
+        // FIXME: year::max() + year::min()
+        static constexpr sys_seconds _Min_seconds{sys_days{year{-32767} / January / 1}};
+        static constexpr sys_seconds _Max_seconds{sys_days{year{32767} / December / 31}};
+
     private:
+        template <class _Duration>
+        _NODISCARD sys_info _Get_info(const _Duration& _Dur) const {
+            using _Internal_duration = duration<__std_tzdb_epoch_milli, std::milli>;
```
[the remark this answers is not in the document]
Comment (d-winsor) [reply stl/inc/chrono:2379]: This was just accidental on my part.

File: stl/inc/chrono [hunk 21]
```diff
@@ -2229,7 +2270,140 @@ namespace chrono {
...
+            return _Info;
+        }
+
+        template <class _Duration>
+        _NODISCARD sys_time<common_type_t<_Duration, seconds>> to_sys(const local_time<_Duration>& _Local) const {
+            auto _Info = get_info(_Local);
+            if (_Info.result == local_info::nonexistent) {
+                throw nonexistent_local_time(_Local, _Info);
+            } else if (_Info.result == local_info::ambiguous) {
+                throw ambiguous_local_time(_Local, _Info);
```
[the remark this answers is not in the document]
Comment (d-winsor) [reply stl/inc/chrono:2346]: Braces will result in error when used with _THROW macro unless the exception is surround by additional parentheses: `_THROW((nonexistent_local_time{_Local, _Info}));`. Which would be preferred here?

File: stl/src/tzdb.cpp [hunk 22]
```diff
@@ -154,56 +209,52 @@ namespace {
...
-        return _Link_buf;
+        return _Data;
     }
 
-    _NODISCARD _STD unique_ptr<char16_t[]> _Get_default_timezone(
-        int32_t& _Result_len, __std_tzdb_error& _Err) noexcept {
-        constexpr int32_t _Name_buf_len = 32;
-        _STD unique_ptr<char16_t[]> _Name_buf{new (_STD nothrow) char16_t[_Name_buf_len]};
-        if (_Name_buf == nullptr) {
+    using _Icu_get_string_fn = std::function<int32_t(UChar* _Result, int32_t _Result_capacity, UErrorCode* _UErr)>;
```
[the remark this answers is not in the document]
Comment (d-winsor) [reply stl/src/tzdb.cpp:239]: Removed all together.

File: stl/src/tzdb.cpp [hunk 23]
```diff
@@ -213,17 +264,62 @@ namespace {
...
+        const auto _Display_type =
+            _Is_daylight ? UCalendarDisplayNameType::UCAL_SHORT_DST : UCalendarDisplayNameType::UCAL_SHORT_STANDARD;
+        _Icu_get_string_fn _Icu_fn = [_Cal, _Display_type](
+                                         UChar* _Result, int32_t _Result_capacity, UErrorCode* _UErr) {
+            return __icu_ucal_getTimeZoneDisplayName(_Cal, _Display_type, nullptr, _Result, _Result_capacity, _UErr);
+        };
+        return _Get_icu_string_impl(_Icu_fn, 12, _Result_len, _Err);
+    }
+
+    _STD unique_ptr<UCalendar, decltype(&__icu_ucal_close)> _Get_cal(
```
[the remark this answers is not in the document]
Comment (d-winsor) [reply stl/src/tzdb.cpp:298]: This logic is repeated over the ABI layer for `__std_tzdb_get_time_zones`/`__std_tzdb_get_current_zone`/`__std_tzdb_get_sys_info`/`__std_tzdb_get_reg_leap_seconds`. Should we create a templated deleter in `xtzdb.h`?

File: stl/inc/chrono [hunk 24]
```diff
@@ -2229,7 +2270,140 @@ namespace chrono {
...
+            }
+
+            return sys_time<common_type_t<_Duration, seconds>>{_Local.time_since_epoch()} - _Info.first.offset;
+        }
+
+        template <class _Duration>
+        _NODISCARD sys_time<common_type_t<_Duration, seconds>> to_sys(
+            const local_time<_Duration>& _Local, choose _Choose) const {
+            auto _Info   = get_info(_Local);
+            auto _Result = sys_time<_Duration>{_Local.time_since_epoch()};
```
[the remark this answers is not in the document]
Comment (d-winsor) [reply stl/inc/chrono:2356]: This var was unused, removed.

**Review by mnatsuhara — APPROVED** [review]
All I found were nitpicks!  I'm happy to merge this as is and roll these small changes into the follow up PR for `zoned_time`!

File: stl/inc/chrono [hunk 25]
```diff
@@ -2229,7 +2270,144 @@ namespace chrono {
...
+        template <class _Duration>
+        _NODISCARD sys_time<common_type_t<_Duration, seconds>> to_sys(const local_time<_Duration>& _Local) const {
+            const auto _Info = get_info(_Local);
+            if (_Info.result == local_info::nonexistent) {
+                _THROW(nonexistent_local_time(_Local, _Info));
+            } else if (_Info.result == local_info::ambiguous) {
+                _THROW(ambiguous_local_time(_Local, _Info));
+            }
+
+            return sys_time<common_type_t<_Duration, seconds>>{_Local.time_since_epoch()} - _Info.first.offset;
```
Comment (statementreply) [inline stl/inc/chrono:2349]:
```suggestion
            return sys_time<common_type_t<_Duration, seconds>>{_Local.time_since_epoch() - _Info.first.offset};
```

I'm okay with it as is, but `sys_time{local.time_since_epoch - offset}` sounds more logically "correct" to me: `offset` is the difference between the epoches of `local_time` and `sys_time`, not the difference between the time points themselves (they represent the same physical time point).

Same below.

File: stl/inc/chrono [hunk 26]
```diff
@@ -2229,7 +2270,144 @@ namespace chrono {
...
+            const auto _Offset = (_Info.result == local_info::unique || _Choose == choose::earliest)
+                                   ? _Info.first.offset
+                                   : _Info.second.offset;
+            return sys_time<common_type_t<_Duration, seconds>>{_Local.time_since_epoch()} - _Offset;
+        }
+
+        template <class _Duration>
+        _NODISCARD local_time<common_type_t<_Duration, seconds>> to_local(const sys_time<_Duration>& _Sys) const {
+            const auto _Info = get_info(_Sys);
+            return local_time<common_type_t<_Duration, seconds>>{(_Sys + _Info.offset).time_since_epoch()};
```
Comment (statementreply) [inline stl/inc/chrono:2369]:
```suggestion
            return local_time<common_type_t<_Duration, seconds>>{_Sys.time_since_epoch() + _Info.offset};
```

Also more logically "correct", okay with it as is.

File: stl/inc/chrono (hunk 21 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/chrono]: Extra parens are fine, we use those with macros when necessary. (We should probably consider variadic macros, I'll add a todo for that)

File: stl/src/tzdb.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/src/tzdb.cpp]: That would be great!

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 27]
```diff
@@ -80,21 +95,294 @@ void timezone_names_test() {
...
+
+void validate_get_local_info(const time_zone* tz, const pair<Transition, Transition>& transition, int result) {
+    sys_info default_info{};
+    sys_info first  = tz->get_info(transition.first.begin());
+    sys_info second = tz->get_info(transition.second.begin());
+
+    // Get the local time for the beginning of the ambiguous/nonexistent section
+    const auto danger_begin = get_danger_begin(transition.first, transition.second);
+    const auto danger_end   = get_danger_end(transition.first, transition.second);
+    (void) result;
```
[the remark this answers is not in the document]
Comment (d-winsor) [reply tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:250]: must have been testing stuff by commenting out lines. Will fix.

File: tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp [hunk 28]
```diff
@@ -80,21 +95,294 @@ void timezone_names_test() {
...
+        const auto& to_standard = Day_to_Std.first.end();
+        assert(tz->to_local(to_standard) == Day_to_Std.second.local_begin());
+        assert(tz->to_local(to_standard + minutes{30}) == Day_to_Std.second.local_begin() + minutes{30});
+        assert(tz->to_local(to_standard - minutes{30}) == Day_to_Std.first.local_end() - minutes{30});
+
+        const auto& to_daylight = Std_to_Day.first.end();
+        assert(tz->to_local(to_daylight) == Std_to_Day.second.local_begin());
+        assert(tz->to_local(to_daylight + minutes{30}) == Std_to_Day.second.local_begin() + minutes{30});
+        assert(tz->to_local(to_daylight - minutes{30}) == Std_to_Day.first.local_end() - minutes{30});
+    }
```
[the remark this answers is not in the document]
Comment (d-winsor) [reply tests/std/tests/P0355R7_calendars_and_time_zones_time_zones/test.cpp:208]: This was an artifact of this being restructured. I think I will leave for now, makes asserts clearer if one fails.

File: stl/inc/chrono [hunk 29]
```diff
@@ -2229,7 +2270,144 @@ namespace chrono {
...
+        }
+
+        // FIXME: year::max() + year::min()
+        static constexpr sys_seconds _Min_seconds{sys_days{year{-32767} / January / 1}};
+        static constexpr sys_seconds _Max_seconds{sys_days{year{32767} / December / 31}};
```
[the remark this answers is not in the document]
Comment (d-winsor) [reply stl/inc/chrono]: I went with  your suggestion `-seconds{1}` to ensure `sys_info::end` would be a valid time.

File: stl/inc/chrono (hunk 21 above)
[the remark this answers is not in the document]
Comment (d-winsor) [reply stl/inc/chrono]: I'll sneak this through in the next `zoned_time` PR.

File: stl/src/tzdb.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (d-winsor) [reply stl/src/tzdb.cpp]: I'll sneak this through in the next `zoned_time` PR.

**Review by mnatsuhara — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]
